test/precompile_tests.jl

```diff
@@ -0,0 +1,99 @@
+@testitem "Precompile" begin
+
+using GPUCompiler
+pushfirst!(LOAD_PATH, joinpath(@__DIR__, "Native"))
+
+import Native
+
+function precompile_test_harness(@nospecialize(f), testset::String)
+    @testset "$testset" begin
+        precompile_test_harness(f, true)
+    end
+end
+function precompile_test_harness(@nospecialize(f), separate::Bool)
+    load_path = mktempdir()
+    load_cache_path = separate ? mktempdir() : load_path
+    try
+        pushfirst!(LOAD_PATH, load_path)
+        pushfirst!(DEPOT_PATH, load_cache_path)
+        f(load_path)
+    finally
+        try
+            rm(load_path, force=true, recursive=true)
+        catch err
+            @show err
+        end
+        if separate
+            try
+                rm(load_cache_path, force=true, recursive=true)
+            catch err
+                @show err
+            end
+        end
+        filter!((≠)(load_path), LOAD_PATH)
+        separate && filter!((≠)(load_cache_path), DEPOT_PATH)
+    end
+    nothing
+end
+
+function check_presence(mi, token)
+    found = false
+    ci = isdefined(mi, :cache) ? mi.cache : nothing
+    while ci !== nothing
+        if ci.owner === token && ci.max_world == typemax(UInt)
+            found = true
+            break
+        end
+        ci = isdefined(ci, :next) ? ci.next : nothing
+    end
+    return found
+end
+
+precompile_test_harness("Inference caching") do load_path
+    write(joinpath(load_path, "InferenceCaching.jl"), :(module InferenceCaching
+        import Native
+        import GPUCompiler
+
+        function kernel()
+            return
+        end
+
+        let
+            job, _ = Native.create_job(kernel, ())
+            GPUCompiler.code_typed(job)
+        end
+        
+        # identity is foreign
+        # Maybe https://github.com/JuliaLang/julia/pull/49391
+        job, _ = Native.create_job(identity, (Int,))
```

Comment:
@timholy can you remind me of the rules around caching of foreign code-instances,

I expected this to work, but maybe I need to dive into how these CodeInstances get queued.